// rs.user/services/index.js
var rs = require('../misc/rs');
var path = require('path');
var schema = require('../misc/schema');
var fs = require('fs');

var service = rs.classy.define({
    forceInstance: true,
    init: function (context) {
        if(rs.config.has('rs.user.db.conn'))
            context.conn = rs.config.get('rs.user.db.conn');

        var carriers = rs.config.has('rs.raven.carriers') ? rs.config.get('rs.raven.carriers'): null;
        var raven = new rs.raven({carriers: carriers});

        //Loading email templates
        var tpl_default = require('../misc/templates/email');
        var tpl_custom_path;
        if(rs.config.has('rs.user.templates.email')) {
            tpl_custom_path = rs.config.get('rs.user.templates.email');
        }
        var tpl_custom;

        try{
            tpl_custom = require(tpl_custom_path);
        } catch (e){
            rs.logw('rs.user.templates - ' + e.message);
        }

        var templates = {
            email: rs.apply(tpl_default, tpl_custom)
        };

        //console.log(templates);;

        var ctx = {
            templates: templates,
            raven : raven,
            schema: rs.data.model.cleanSchema(schema, { addDefaults: true }),
            client: rs.data.client({type: 'mssql', conn: context.conn})
        };

        this.user = require('./user')(ctx);
        this.auth = require('./auth')(ctx);
        this.comm = require('./comm')(ctx)
    }
});

module.exports = service;
