// rs.raven/src/carriers/messenger.js
var rs = require('../misc/rs');
var _base = require('./_base');
var request = require('request');

var tpm_attachment = {
    attachment: {
        type: "image",
        payload: {
            url: ''
        }
    }
};
/**
 * @classdesc
 * @constructor
 */
var service = rs.classy.define(/** @lends rs.raven.services# */{
    extend: _base,
    init: function (c) {
        rs.apply(this, c);
        this.url = "https://graph.facebook.com/v2.6/me/messages?access_token=" + c.config.access_token;
        this.callSuper()
    },
    send: function (msg, options) {
        if(typeof(msg.body)==='string')
            msg.body = {text: msg.body};

        if(msg.body.image){
            tpm_attachment.attachment.payload.url = msg.body.image;
            msg.body = tpm_attachment;
        } else if(msg.body.video) {
            tpm_attachment.attachment.type = 'video';
            tpm_attachment.attachment.payload.url = msg.body.video;
            msg.body = tpm_attachment;
        } else if(msg.body.file) {
            tpm_attachment.attachment.type = 'file';
            tpm_attachment.attachment.payload.url = msg.body.file;
            msg.body = tpm_attachment;
        } else if(msg.body.audio) {
            tpm_attachment.attachment.type = 'audio';
            tpm_attachment.attachment.payload.url = msg.body.audio;
            msg.body = tpm_attachment;
        }

        request.post({
            url: this.url,
            headers: {'content-type':'application/json'},
            json: {
                recipient: {
                    id: msg.to
                },
                message: msg.body
            }
        }, function(error, httpResponse, content){
            if(error){
                if(options.fail)
                    return options.fail(error);
            }
            if(options.success)
                options.success(content);
        });
    }
});

module.exports = service;
